### ISSUE TITLE
API Server: Use workers/clusters by default

### ISSUE THREAD

**Issue by cannikin** [redwoodjs/redwood#3036]
I just learned that our api server is basically single-threaded and every request to the api in a serverfull environment blocks everything (probably not *everything*, but we use a lot of async/await).

I just assumed that Express already did this (isn't it *THE* Node web server?), but apparently not!

The number of workers should be configurable, maybe defaulting to number of cpus * 2 (most processes won't use 100% of the CPU so you'll generally see a recommendation to use 2 worker processes per CPU). 

This is pretty essential for anyone doing serverfull deployment—you should be able to max out the CPUs on a single server before having to worry about adding multiple servers and load balancing them. Right now having a 2/4/8 CPU machine may as well only have 1 because that's all that's ever going to get used. :(

**Comment by dac09** [comment]
I think we get it for free if we setup `rw serve` to run with pm2 in cluster mode, with no code changes. 

https://pm2.keymetrics.io/docs/usage/cluster-mode/

**Comment by dac09** [comment]
Ooo guess what @cannikin - pm2 also gives us zero downtime restarts

**Comment by cannikin** [comment]
Sounds like a win-win! Who wants to be my best friend and add that? 😃

I'm wary because any time I've messed with processes in basically any OS or language I've ever used I wanted to jump out a window, and this being the Javascript ecosystem I imagine it's 2x as messy. But don't let me talk you out of it. 😅

**Comment by dac09** [comment]
## Recommended path to solution
Setup pm2 to run the web server, with cluster-mode on. This would give us all the benefit, without changing any of the api-server code, and also maximise compatibility.

**Comment by jeliasson** [comment]
By adding pm2, aren't we adding more project dependencies and cluttering down the projects with additional files? 

For me the natural solution would be to add more instances and scale horizontal, or use NodeJS's [cluster api](https://nodejs.org/api/cluster.html).

**Comment by thedavidprice** [comment]
Oh, cluster api looks interesting.

I'm mixed on PM2 being _the_ solution, but I do think it would make the default 100x better. And the [package](https://www.npmjs.com/package/pm2) doesn't look terrible given that it would be for the api-server for containerization use.

**Comment by cannikin** [comment]
@jeliasson My reason for wanting pm2 is just to keep my web/api processes monitored to make sure they keep running. If they die for whatever reason, I'm screwed.

pm2 apparently also supports hot restarts—right now I have manually quit both processes and restart them on each deploy.

I'm running on regular ol' servers by the way, no lambdas, no containers, just EC2 instances. I'm serving both directly to the client, with nginx sitting in front of the web side to proxy 80/443 to port 8910 and add some cache headers. Works great!

**Comment by jeliasson** [comment]
@cannikin PM2 is great but I don't think it's the right approach as a default `rw serve`. Approaching this from a container perspective, here are a few things that comes to mind;

- A container (a entrypoint process) that never dies, because something in between makes sure of it, could throw off native healthchecks. If a process don't die, how do orchestrating engines (Docker Swarm, Kubernetes, etc) know when something is wrong?
- Logs from a process that spawns sub-processes don't consolidate that well. Not a dealbreaker, but adds another layer.
- More depdendencies results in longer build times and larger images. However, I'm guessing it's minuscule in the grand schema of things.

I might be on the wrong on above, but considering this a somewhat a edge case, maybe a `rw setup pm2` along with `rw pm2` (a wrapper, just like `rw prisma`) would be sufficient?

In the meantime, here is a PM2 configuration you can throw in locally to get you started 😊

**ecosystem.config.js**
```js
module.exports = {
  apps: [
    {
      name: 'api',
      script: 'node_modules/.bin/rw',
      args: 'serve api',
      instances: 2,
      exec_mode: 'cluster',
    },
    {
      name: 'web',
      script: 'node_modules/.bin/rw',
      args: 'serve web',
      instances: 2,
      exec_mode: 'cluster',
    },
  ],
}
```

**Fly 🚀**
```bash
yarn add -D -W pm2
yarn rw build
yarn pm2 start
yarn pm2 logs
```

**Comment by dac09** [comment]
This is really great @jeliasson - this is one of those issues I can never get to! Appreciate your input on this!

100% agreed with you personally on just pm2 isn't enough, in a serious application. 

I think the idea behind this is that we support all levels of applications - people who want to easily host resilient smaller projects, as well as leaving enough room for large scale applications. 

Given this would you say a setup command is better option? The alternative being having pm2 built into the serve command (would this interfere with using more sophisticated tools?) 

Very keen to hear your thoughts on this - definitely more your area of expertise.

**Comment by jeliasson** [comment]
> Given this would you say a setup command is better option? The alternative being having pm2 built into the serve command (would this interfere with using more sophisticated tools?)

@dac09 Maybe `rw setup pm2` (that scaffolds a ecosystem.config.js file and brings in pm2 as dependency) along with e.g. `rw serve pm2 [...]` as wrapper, which I think it what you meant by the alternative, would be better. :)

In the end, I think we're doing us a favor if we have a 'clean entry point' (`yarn rw serve [api/web]`) that honors process exit codes and is somewhat predictable in it's output. Other than that, I would love to see being pm2 added.

**Comment by cannikin** [comment]
@jeliasson Next part of the puzzle: how to build the web side for a new deploy while the `dist` folder is being served to the live site—currently the live assets will change as the build progresses because it's overriding the files in there. The way that Rails deploys typically handle this is by checking out a new copy of the repo into a folder named with a timestamp, and buliding in there. nginx is configured to serve static assets from a directory called `current` which is actually just a symlink to the built files in the directory-named repo that's currently live. So you might start out like this:

```
/var/www/current  ->  /var/www/20211003120000/web/dist
```

On a new deploy, this happens:

```
/var/www/current  ->  /var/www/20211002120000/web/dist
                      /var/www/20211003183000  <- code checked out here, build process started
```

And then when the build is done, assuming no errors, the last step is to change where the symlink is pointing:

```
                       /var/www/20211002120000  <- keep this around for a little while in case a rollback is needed, but eventually delete
 /var/www/current  ->  /var/www/20211003183000/web/dist
```

Right now I'm using `web serve` but I should be able to just have nginx serve the `dist` directory manually and this symlink magic should be good to go.

The last part of the puzzle is hot reloading the api-side without losing any connections. Is there a `pm2 reload` command, or similar?

**Comment by dac09** [comment]
Thanks for getting the PR started @jeliasson - I'll continue the rest of the conversation there.

But I just wanted to confirm that "0 downtime" reloads are possible with pm2 (just tried it!), with `yarn pm2 reload`. Seamlessly switches the servers over, killing the old processes!

**Comment by peterp** [comment]
Just want to point out that pm2 uses the agpl license. Since our target audience are startups whom will want to be in the enterprise we should not be adding this package.

**Comment by jeliasson** [comment]
@peterp Good catch! I'm on the fence of adding pm2 at all, outlining my concerns from a serverful perspective [here](https://github.com/redwoodjs/redwood/issues/3036#issuecomment-932639663), but I'm not entirely sure about the scope. This issue seem to indicate that it would, initially, be part of the default `yarn rw serve` of Redwood. That is a 🚩flag to me. However, I can definitely see being used as part of a `yarn rw setup pm2`. Maybe with clear information on it's licensing. 

Trying to not step on anyone's toes, but why would pm2 be considered as a default package for Redwood to begin with? ✌️

**Comment by jtoar** [comment]
@peterp @jeliasson we discussed this briefly at a meeting. The team's answer was: agreed that baking it into `yarn rw serve` isn't a good idea, but having it as a part of a setup command via `yarn rw setup pm2` seems ok IF we include some kind of prompt that users have to say yes/no to, just so that they're explicitly aware that they're including a package with an agpl license. Any further thoughts? I can always queue it back up again for discussion!
